### PR TITLE
Add automated benchmarking

### PR DESCRIPTION
**Pull request by peytondmurray**
This PR adds automated benchmarking on GCP via Cirun:

- Added `.cirun.yml` configuration. Seems to be working well already :tada: 
- Made a benchmarking workflow that pushes results to `master`. When the docs build, the asv results are rendered into static HTML. A link to these is added to the front page of the docs
- The benchmarking install code has been _greatly_ simplified
- Reduced the number of versions in the `delete_version` benchmark; it was taking way too long

Closes https://github.com/deshaw/versioned-hdf5/issues/343.

This was tested with a pure python version of `build_data_dict`, as the current one is broken on CI and my local machine: https://github.com/peytondmurray/versioned-hdf5/actions/runs/11391825172

### CODE REVIEW

File: benchmarks_install.py [hunk 1]
```diff
@@ -7,77 +8,57 @@
...
-    sys.stdout.flush()
+logger = logging.getLogger(__name__)
 
-def run(command, *args, **kwargs):
-    print(' '.join(command))
-    kwargs.setdefault('check', True)
-    return subprocess.run(command, *args, **kwargs)
+# The first commit in versioned-hdf5 that is not compatible with ndindex 1.5;
+# this was released in versioned-hdf5 1.2.2
+ndindex_1_5_pin_version = '1.2.2'
```
Comment (crusaderky) [inline benchmarks_install.py:19]: ndindex 1.6 is almost 3 years old; versioned_hdf5 is 1.2.2 is almost 4 years old (which makes no sense to me given the comment above). 

Does anyone even care about versions this old? Shouldn't we just stat from a sane minimum version e.g. whatever was published in the last 2 years?

IMHO I would delete this whole script.

File: docs/index.rst [hunk 2]
```diff
@@ -18,8 +18,7 @@ Versioned HDF5 provides a versioned abstraction on top of `h5py <https://www.h5p
 Benchmarks
 ----------
 
-Benchmarks for Versioned HDF5 are published `here
-<https://deshaw.github.io/versioned-hdf5/benchmarks/index.html>`_.
+Benchmarks for Versioned HDF5 are published `here <../benchmarks/index.html>`_.
```
Comment (crusaderky) [inline docs/index.rst:21]: I can't see anything at https://peytondmurray.github.io/versioned-hdf5/benchmarks/index.html, is the output uploaded anywhere?

**Comment by crusaderky** [comment]
@peytondmurray I think it would be incredibly useful to be able to rerun asv on a controlled environment (GCP) on a PR before it gets merged and compare the results. Do I understand correctly however that this PR exclusively impacts master? Would it be a sensible followup to add the functionality to do a comparative benchmark on an open PR (maybe triggered by a github PR tag)?

File: benchmarks_install.py (hunk 1 above)
Comment (peytondmurray) [reply benchmarks_install.py:19]: Sure, I'm happy to delete this entirely.

File: docs/index.rst (hunk 2 above)
Comment (peytondmurray) [reply docs/index.rst:21]: This was replaced in a more recent commit on my `master` branch.

**Comment by peytondmurray** [comment]
> I think it would be incredibly useful to be able to rerun asv on a controlled environment (GCP) on a PR before it gets merged and compare the results

Sure, this is definitely possible - especially if you're fine with just a table of benchmark results as output. I can follow up in another PR with a bot that posts benchmarks in a comment, if that makes the most sense.

File: .asv/results/cirun-peytondmurray--versioned-hdf5-ffd0030/52aca7e9-virtualenv-py3.11.json [hunk 3]
```diff
@@ -0,0 +1 @@
...[0.075741], [0.080987], [0.076092], [0.077583], [1], [10], [[0.07593489499998896, 0.07597559800001363, 0.07574061299999357, 0.07728835799997569, 0.08098718000002236, 0.07768118900003174, 0.07661628300002121, 0.07801443899995775, 0.07701543000007405, 0.07643924400008473]]], "resize.time_resize_hdf5": [[0.6056949905000124], [], "4e9ed3924a20c3eb88bff7199efce32a2f66cf82247e481040b00f67b6e21ed9", 1729194584030, 3.6298, [0.58675], [0.61544], [0.60392], [0.61032], [1], [10], [[0.5867515289999687, 0.6070298679999837, 0.6058430210000552, 0.6121450240000286, 0.6037039770000092, 0.5888744210000141, 0.6045584729999973, 0.6154410769999004, 0.6055469599999697, 0.6114153470000474]]], "versionedhdf5file.TimeDatetimeAccess.time_version_by_datetime": [[0.031923934999980474], [], "68650944530784b3fda1ba4817d8e831fda665cd7ff0a32c4e20e92aaac32816", 1729194587660, 3.0892, [0.03106], [0.035014], [0.031699], [0.032789], [1], [10], [[0.031059697999978653, 0.031718804000036016, 0.03201599400000532, 0.03303554100000383, 0.035013658000025316, 0.03183187599995563, 0.03169215900004474, 0.03205045900006098, 0.03160499600005551, 0.03330933200004438]]]}, "durations": {"<build>": 23.980255842208862}, "version": 2}
```
[the remark this answers is not in the document]
Comment (peytondmurray) [reply .asv/results/cirun-peytondmurray--versioned-hdf5-ffd0030/52aca7e9-virtualenv-py3.11.json:1]: Strictly speaking we don't need these results, but we do need these files to be tracked in git. I don't think they'll do any harm here, though?

**Comment by peytondmurray** [comment]
Here's the latest version, I merged this PR into my own fork's `master` to trigger the build: https://peytondmurray.github.io/versioned-hdf5/benchmarks/index.html

I think the graphs aren't displaying correctly because there's only one data point (the latest commit comparison), but it could be that `master~1` doesn't have the `H5Close` commit, so the benchmark failed. Anyway that's not terribly relevant as the machinery there is being changed out soon anyway.

**Review by crusaderky — APPROVED** [review]

**Comment by crusaderky** [comment]
@peytondmurray it's failing in master:
1. still the dcpl.id not being found, and
2. something about git permissions when you try pushing

https://github.com/deshaw/versioned-hdf5/actions/runs/11476804181/job/31937575352

**Comment by crusaderky** [comment]
You're installing libhdf5 from conda and h5py from pip. Could you try moving h5py to conda too?

**Comment by peytondmurray** [comment]
Yep, will do now.

**Pull request merged**